### DISCUSSION TITLE
Efficient way to create a list of Point.

### DISCUSSION THREAD

**Discussion by tianzhuqiao** [shapely/shapely#2117] [Q&A] [answered]
## Please note
I want to create a list of `Point` (e.g.,  10k) to check if they are in a polygon. I tried the following ways, looks like the most of the time is to generate the Point list (e.g., 75%). What will be the most efficient way to do that? Thanks!
```python
    p = [Point(x, y) for x, y in zip(xq, yq)]
    df = pd.DataFrame({'x':xq, 'y':yq, 'coords': p})
```
```python
    df = pd.DataFrame({'x':xq, 'y':yq})
    df['coords'] = list(zip(df['x'],df['y']))
    df['coords'] = df['coords'].apply(Point)
```

**Comment by mwtoews** [comment, marked as answer]
Assuming you are using Shapely 2.x, you can use the [`points`](https://shapely.readthedocs.io/en/stable/reference/shapely.points.html#shapely.points) function, which is vectorized to return numpy array types.

```python
import numpy as np
import shapely

rng = np.random.default_rng(123)
n = 10_000
xq = rng.random(n)
yq = rng.random(n)
pts = shapely.points(xq, yq)
```
and when you get to do the point-in-polygon check, also use vectorized functions like `contains`:
```python
poly = shapely.from_wkt("POLYGON((0 0, 0 0.1, 0.1 0.1, 0 0))")
pt_in_poly = shapely.contains(poly, pts)
# or poly.contains(pts)
pt_in_poly.sum()  # 60
```

**Reply by tianzhuqiao** [reply]
Thank you so much, @mwtoews! Just checked, the run time drops significantly (750s to 16s).

**Comment by tianzhuqiao** [comment]
I am trying to create a "equivalent" `inpolygon` function in matlab with python. With the vectorized function in shapely, it is much fast. But looks like it is still 3~4 times slower than matlab. The following is the current code, is it the right way to do it?

https://www.mathworks.com/help/matlab/ref/inpolygon.html

```python
def inpolygon(xq, yq, xv, yv):
    polygon = Polygon(list(zip(xv, yv)))
    if not polygon.is_valid:
        polygon = polygon.buffer(0)

    pts = shapely.points(xq, yq)
    return polygon.touches(pts) | polygon.contains(pts)
```

And this is the result with [line_profile](https://github.com/pyutils/line_profiler). Is there some way to improve it? Thank you so much!
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   210                                           @profile
   211                                           def inpolygon(xq, yq, xv, yv):
   212 
   213      7208    4937984.6    685.1     17.4      polygon = Polygon(list(zip(xv, yv)))
   214      7208     409624.9     56.8      1.4      if not polygon.is_valid:
   215                                                   polygon = polygon.buffer(0)
   216                                           
   217      7208    9856676.3   1367.5     34.7      pts = shapely.points(xq, yq)
   218      7208   13222215.3   1834.4     46.5      return polygon.touches(pts) | polygon.contains(pts)

```

**Reply by theroggy** [reply]
You could try:
- replacing `list(zip(xv, yv))` by `np.column_stack((xv, yv))`
- replacing `polygon.touches(pts) | polygon.contains(pts)` by `polygon.intersects(pts)`, that should be faster. 
- depending on the version of GEOS being used it can also help to prepare your polygon before the intersects with `shapely.prepare(polygon)`.

**Reply by tianzhuqiao** [reply]
Thank you so much, @theroggy. It really helps. For the testcase, originally it  (the above `inpolygon` function) takes 28 seconds,
1. with your 1st suggestion above (`np.column_stack((xv, yv))`), it reduces to 24 seconds
2. with your 1st and 3rd suggestions above, it reduces to 18 seconds
3. with all above 3 changes, it reduces to 11 seconds.

**Reply by theroggy** [reply]
If it is still taking 11 seconds, this can most likely, depending on your data and/or if your code can be structured a bit differently be further optimised by using a spatial index...

**Reply by tianzhuqiao** [reply]
Got it. Right, with all your changes, it is close to matlab now. Thank you so much!
